File: benchmarks/benchmark_mlp.jl
"""
MLP Training Benchmark (Julia)
Equivalent to benchmark_mlp.py using Lux, Enzyme, and Reactant on the MLX backend.

Usage:
  julia +1.11 benchmarks/benchmark_mlp.jl
"""

using Libdl, Reactant, Lux, Enzyme, Random, Statistics

# ── MLX Plugin Setup ──────────────────────────────────────────────────────────

function load_mlx_plugin!()
    mlx_lib = "/Users/thomas/miniforge3/envs/jax/lib/python3.13/site-packages/mlx/lib/libmlx.dylib"
    Libdl.dlopen(mlx_lib, Libdl.RTLD_NOW | Libdl.RTLD_GLOBAL)

    plugin_path = joinpath(@__DIR__, "..", "src", "jax_mlx", "libmlx_pjrt_plugin.dylib")
    if !isfile(plugin_path)
        plugin_path = joinpath(@__DIR__, "..", "build", "libmlx_pjrt_plugin.dylib")
    end
    @assert isfile(plugin_path) "Cannot find libmlx_pjrt_plugin.dylib at $plugin_path"

    mlx_client_ptr = Reactant.XLA.PJRT.MakeClientUsingPluginAPI(plugin_path, "mlx", "MLX")
    mlx_client = Reactant.XLA.PJRT.Client(mlx_client_ptr)
    Reactant.XLA.global_backend_state.clients["mlx"] = mlx_client
    Reactant.XLA.set_default_backend(mlx_client)
    println("MLX plugin loaded ✓")
end

# ── MLP Model ─────────────────────────────────────────────────────────────────

function SimpleMLP()
    return Chain(
        Dense(2048, 8192, relu),
        Dense(8192, 4096, relu),
        Dense(4096, 4096, relu),
        Dense(4096, 2048, relu),
        Dense(2048, 1024, relu),
        Dense(1024, 10),
    )
end

# ── Loss Function ─────────────────────────────────────────────────────────────

function cross_entropy_loss(model, ps, st, x, y)
    logits, st_ = model(x, ps, st)
    log_probs = logits .- Lux.logsumexp(logits; dims=1)
    classes = reshape(1:10, 10, 1)
    labels = reshape(y, 1, :)
    one_hot = Float32.(classes .== labels)
    return -sum(one_hot .* log_probs) / Float32(size(logits, 2))
end

# ── Train Step ────────────────────────────────────────────────────────────────

function train_step(model, ps, st, x, y, lr)
    dps = Enzyme.make_zero(ps)
    _, loss = Enzyme.autodiff(
        Enzyme.ReverseWithPrimal,
        cross_entropy_loss,
        Active,
        Const(model),
        Duplicated(ps, dps),
        Const(st),
        Const(x),
        Const(y),
    )
    ps_new = Lux.recursive_map((p, g) -> p .- lr .* g, ps, dps)
    return ps_new, loss
end

# ── Benchmark ─────────────────────────────────────────────────────────────────

function benchmark_mlp(;
    batch_size::Int=2048,
    input_dim::Int=2048,
    num_warmup::Int=30,
    num_runs::Int=20,
    lr::Float32=0.01f0,
)
    println("MLP Training Benchmark (Julia/Lux/Reactant)")
    println("Config: batch_size=$batch_size, input_dim=$input_dim")
    println("Model: Dense(1024→8192→4096→4096→2048→1024→10) + ReLU")
    println("Task: Forward + Backward pass with SGD\n")

    load_mlx_plugin!()

    model = SimpleMLP()
    rng = Random.Xoshiro(0)
    ps, st = Lux.setup(rng, model)
    ps = ps |> Lux.f32

    x = randn(Float32, input_dim, batch_size)
    y = rand(1:10, batch_size)

    x_r = Reactant.to_rarray(x)
    y_r = Reactant.to_rarray(y)
    ps_r = Reactant.to_rarray(ps)
    st_r = Reactant.to_rarray(st)

    println("Compiling train step...")
    t_compile_start = time()
    compiled_step = @compile train_step(model, ps_r, st_r, x_r, y_r, lr)
    t_compile = time() - t_compile_start
    println("Compilation done in $(round(t_compile; digits=1))s ✓\n")

    println("Warming up ($num_warmup steps)...")
    for i in 1:num_warmup
        ps_r, loss = compiled_step(model, ps_r, st_r, x_r, y_r, lr)
    end
    println("Warmup done ✓\n")

    println("="^50)
    println("Timed runs")
    println("="^50)
    times = Float64[]
    for i in 1:num_runs
        t0 = time_ns()
        ps_r, loss = compiled_step(model, ps_r, st_r, x_r, y_r, lr)
        loss_val = Float64(loss)
        t1 = time_ns()
        dt_ms = (t1 - t0) / 1e6
        push!(times, dt_ms)
        println("  Run $i: $(round(dt_ms; digits=2))ms (loss: $(round(loss_val; digits=4)))")
    end

    steady = times[2:end]
    mean_t = mean(steady)
    std_t = std(steady)
    min_t = minimum(steady)
    println("\n  Mean: $(round(mean_t; digits=2))ms ± $(round(std_t; digits=2))ms  Min: $(round(min_t; digits=2))ms  (excl. first run)")

    println("\n" * "="^50)
    println("SUMMARY")
    println("="^50)
    println("Julia/Lux/Reactant/MLX: $(round(mean_t; digits=2))ms ± $(round(std_t; digits=2))ms")
end

benchmark_mlp()

File: benchmarks/benchmark_mlp.py
"""
MLP Training Benchmark: JAX-MLX
Full training loop with forward + backward + SGD.

Usage:
  python benchmarks/benchmark_mlp.py
"""

import os
if "MLX_PJRT_DEBUG" in os.environ:
    del os.environ["MLX_PJRT_DEBUG"]
if "JAX_PLATFORMS" not in os.environ:
    os.environ["JAX_PLATFORMS"] = "cpu,mlx"
elif "cpu" not in os.environ["JAX_PLATFORMS"]:
    os.environ["JAX_PLATFORMS"] = "cpu," + os.environ["JAX_PLATFORMS"]

import jax
import jax.numpy as jnp
from flax import linen as nn
import time
import numpy as np


class MLP(nn.Module):
    """Deep wide MLP for benchmarking."""

    @nn.compact
    def __call__(self, x):
        x = nn.Dense(features=8192)(x)
        x = nn.relu(x)
        x = nn.Dense(features=4096)(x)
        x = nn.relu(x)
        x = nn.Dense(features=4096)(x)
        x = nn.relu(x)
        x = nn.Dense(features=2048)(x)
        x = nn.relu(x)
        x = nn.Dense(features=1024)(x)
        x = nn.relu(x)
        x = nn.Dense(features=10)(x)
        return x


def loss_fn(params, model, x, y):
    """Cross-entropy loss."""
    logits = model.apply(params, x)
    one_hot = jax.nn.one_hot(y, 10)
    return -jnp.mean(jnp.sum(jax.nn.log_softmax(logits) * one_hot, axis=-1))


def create_train_step(model, lr=0.01):
    @jax.jit
    def train_step(params, x, y):
        loss, grads = jax.value_and_grad(lambda p: loss_fn(p, model, x, y))(params)
        params = jax.tree.map(lambda p, g: p - lr * g, params, grads)
        return params, loss
    return train_step


def benchmark_mlx(batch_size=2048, input_dim=2048, num_warmup=30, num_runs=20):
    """Benchmark MLP training on MLX."""
    try:
        mlx = jax.devices("mlx")[0]
    except Exception as e:
        print(f"MLX device not available: {e}")
        return None, None

    model = MLP()

    try:
        cpu = jax.devices("cpu")[0]
    except Exception:
        cpu = mlx

    with jax.default_device(cpu):
        rng = jax.random.PRNGKey(0)
        params = model.init(rng, jnp.ones((batch_size, input_dim)))
        key = jax.random.PRNGKey(42)
        x = jax.random.normal(key, (batch_size, input_dim))
        y = jax.random.randint(key, (batch_size,), 0, 10)

    with jax.default_device(mlx):
        params = jax.device_put(params, mlx)
        x = jax.device_put(x, mlx)
        y = jax.device_put(y, mlx)
        train_step = create_train_step(model)

        # Warmup
        print(f"Warming up ({num_warmup} steps)...")
        for _ in range(num_warmup):
            params, loss = train_step(params, x, y)
            loss.block_until_ready()

        # Timed runs
        print(f"\n{'='*50}")
        print("Timed runs")
        print(f"{'='*50}")
        times = []
        for i in range(num_runs):
            start = time.perf_counter()
            params, loss = train_step(params, x, y)
            loss.block_until_ready()
            end = time.perf_counter()
            times.append(end - start)
            print(f"  Run {i+1}: {times[-1]*1000:.2f}ms (loss: {float(loss):.4f})")

    mean_time = np.mean(times) * 1000
    std_time = np.std(times) * 1000
    min_time = np.min(times) * 1000
    print(f"\n  Mean: {mean_time:.2f}ms ± {std_time:.2f}ms  Min: {min_time:.2f}ms")
    return mean_time, std_time


def main():
    batch_size = 2048
    input_dim = 2048

    print("MLP Training Benchmark (JAX-MLX)")
    print(f"Config: batch_size={batch_size}, input_dim={input_dim}")
    print(f"Model: Dense(1024→8192→4096→4096→2048→1024→10) + ReLU")
    print(f"Task: Forward + Backward pass with SGD\n")

    mean, std = benchmark_mlx(batch_size, input_dim)

    if mean is not None:
        print(f"\n{'='*50}")
        print("SUMMARY")
        print(f"{'='*50}")
        print(f"JAX-MLX: {mean:.2f}ms ± {std:.2f}ms")


if __name__ == "__main__":
    main()
